compiler/rustc_type_ir/src/lib.rs

```diff
@@ -1,4 +1,5 @@
 #![feature(associated_type_defaults)]
+#![feature(auto_traits)]
```

Comment:
So, one point here:

With the recent (renewed) efforts by @compiler-errors to try to abstract things to rustc_type_ir and the *eventual* goal for rust-analyzer to depend on it, this does PR introduce a *very* fundamental dependency on auto traits, for which we don't expect to stabilize soon, if ever.

Reply:
If it weren't for the orphan rule we could make `rustc_type_ir::BoringTraversable` an ordinary (non-auto) trait, then create some other auto trait in `rustc_middle` and blanket impl the former for the latter.

I seem to recall there's an unstable escape hatch from the orphan rule... might that work here/be acceptable?

Reply:
Actually, it should be possible to move the whole specialisation machinery out of the type library and into `rustc_middle` altogether.

Reply:
I think longterm traversal and many of the used types should also be in `rustc_type_ir`, though the `I` param will forever remain.

```
/// The type `T` is trivially traverseable in the interner `Self`
trait BoringTraversable<T>: Interner {}

// in `rustc_middle` using nightly
impl<T: BoringTraversableAuto> BoringTraversable<T> for TyCtxt<'tcx>`
impl !BoriingTraversableAuto for Ty<'tcx> {}
```
I guess: I am confident that the auto trait part of boring traversing can be moved into a nightly `rustc` crate, and believe it to be fine, even in a later PR.

Though ofc, @eggyal if you want to do that move in this PR, that would be even better

Reply:
A broader question for me is: if auto traits are needed here, whether or not they can be kept internal to rustc, should they be being removed? I've been following the discussion on them passively, and I feel this is a situation where you shouldn't have it both ways. Either they're not worth the tradeoffs, in which case the compiler shouldn't be using them, or they are, in which case why can't I (abstract user, not me personally) use them?

Reply:
I've moved the specialisation machinery into `rustc_middle` on my local branch, but it's not without its problems.  In particular, this PR had reached a point where a number of crates upstream of `rustc_middle` were using the derive macros with only a dependency on the type library (most notably including in the type library itself).  With the specialisation machinery now downstream of them, they need an escape hatch (which is then not provable in the type system).

I'm therefore much more inclined to go with your excellent proposal above @lcnr, as it solves the problem perfectly.

Reply:
Actually, I don't think this proposal works after all.  Currently we have:

https://github.com/rust-lang/rust/blob/a416685de1dfd5272c3320e505feb10c9c2a79eb/compiler/rustc_type_ir/src/fold.rs#L277

which would now need to become:

```rust
impl<I: TriviallyTraversable<T>, T> SpecTypeFoldable<I> for &PhantomData<T> {
```

However this additional constraint on the interner [breaks the auto-deref specialisation](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bb49a16b20c77194765be128fe1a7b46).  I guess I'll just stick with moving the specialisation into `rustc_middle`, despite the previously mentioned disadvantages...

Reply:
this happens because `?interner: NewBoringTraversable<T>` is ambiguous, so we use it as it has fewer derefs. You should be able to explicitly specify `I` though, should you?

something like https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a8661757e7e5f73a33cdf1684c349cdb

Reply:
Ah, yes of course :) that works!  Just need to tidy the commits up, rebase and push.  Hopefully tomorrow, but at latest before you get into reviewing over the weekend.

@rustbot author